master/buildbot/reporters/http.py

```diff
@@ -41,11 +41,28 @@ def checkConfig(self, *args, **kwargs):
                          (self.__class__,))
         if not isinstance(kwargs.get('builders'), (type(None), list)):
             config.error("builders must be a list or None")
+        # check that our callbacks are functions
+        if kwargs.get('startCB') and not hasattr(kwargs['startCB'], '__call__'):
+            config.error("start callback must be callable")
+        if kwargs.get('reviewCB') and not hasattr(kwargs['reviewCB'], '__call__'):
+            config.error("review callback must be callable")
+        if kwargs.get('summaryCB') and not hasattr(kwargs['summaryCB'], '__call__'):
+            config.error("summary callback must be callable")
 
     @defer.inlineCallbacks
-    def reconfigService(self, builders=None, **kwargs):
+    def reconfigService(self, builders=None,
+                        startCB=None, startArg=None,
+                        reviewCB=None, reviewArg=None,
+                        summaryCB=None, summaryArg=None,
+                        **kwargs):
         yield service.BuildbotService.reconfigService(self)
         self.builders = builders
+        self.startCB = startCB
+        self.startArg = startArg
+        self.reviewCB = reviewCB
```

Comment:
When you say review (non-gerrit) people think code review. buildbot is not doing code review, it is doing pre-commit. Some say pre-submit checks. Others CI vote.

Gerrit users often confuse ci vote as code review (I plaid guilty myself, as I first introduced reviewCB name in gerrit status)

I know you are working with gerrit, so this is why I suspect you are gerrit biased :)


I clearly don't like callbacks, because they are not as flexible as overridable methods. The main advantage of methods, is that they are passed a 'self', which gives access to master, master.data, so that your callbacks can fetch more data as they like.

Also, we shall factorize what is necessary for the current use cases as we know, not try to factorize for future maybe usecases, because that too often leads to over engineering.
If there is a need for refactor in the longer term, we will refactor at that time.

With your current design, GitlabStatusPush will get a new reviewCB argument, which is completely ignored, and not useful